### ISSUE TITLE
Understanding clusters and cluster aliases

### ISSUE THREAD

**Issue by stankevich** [openark/orchestrator#391]
@shlomi-noach: I'm trying to understand how Orchestrator clusters work, and how they relate to cluster aliases. My goal is to reliably determine a master of each defined cluster from the API using cluster aliases. Here's an example with a single cluster, but this applies to having multiple defined as well.

Relevant configuration:

```
$ orchestrator -version
3.0.6
a77ee001dc95b8dc0476b929164f37b1141f54c1
```

```json
{
  "ApplyMySQLPromotionAfterMasterFailover": true,
  "ClusterNameToAlias": {
    ".*": "my-cluster"
  },
  "DetachLostReplicasAfterMasterFailover": true,
  "DiscoverByShowSlaveHosts": true,
  "HostnameResolveMethod": "none",
  "RecoverIntermediateMasterClusterFilters": [
    ".*"
  ],
  "RecoverMasterClusterFilters": [
    ".*"
  ],
  "RecoveryIgnoreHostnameFilters": [],
}
```

Initial state:

```
$ curl -n -s ${orchestrator_url}/api/topology/my-cluster | jq -r '.Details'
data:3306    [0s,ok,5.7.20-19-log,rw,ROW,>>,GTID]
+ data2:3306 [0s,ok,5.7.20-19-log,ro,ROW,>>,GTID]
+ data3:3306 [0s,ok,5.7.20-19-log,ro,ROW,>>,GTID]
+ data4:3306 [0s,ok,5.7.20-19-log,ro,ROW,>>,GTID]
```

```
$ curl -n -s ${orchestrator_url}/api/clusters | jq -r .
[
  "data:3306"
]
```

```
$ curl -n -s ${orchestrator_url}/api/clusters-info | jq -r .
[
  {
    "HasAutomatedIntermediateMasterRecovery": true,
    "HasAutomatedMasterRecovery": true,
    "HeuristicLag": 0,
    "CountInstances": 4,
    "ClusterDomain": "",
    "ClusterAlias": "my-cluster",
    "ClusterName": "data:3306"
  }
]
```

```
$ curl -n -s ${orchestrator_url}/api/masters | jq -r '.[].Key.Hostname'
data
```

```
$ curl -n -s ${orchestrator_url}/api/master/my-cluster | jq -r '.Key.Hostname'
data
```

Everything looks good, and, by running that last command, I can see that `data` is the master of `my-cluster`. However, this isn't the case if `data` fails and `data2` is promoted:

```
$ curl -n -s ${orchestrator_url}/api/topology/my-cluster | jq -r '.Details'
data:3306 [unknown,invalid,5.7.20-19-log,rw,ROW,>>,GTID]
```

```
$ curl -n -s ${orchestrator_url}/api/clusters | jq -r .
[
  "data:3306",
  "data2:3306"
]
```

```
$ curl -n -s ${orchestrator_url}/api/clusters-info | jq -r .
[
  {
    "HasAutomatedIntermediateMasterRecovery": true,
    "HasAutomatedMasterRecovery": true,
    "HeuristicLag": 0,
    "CountInstances": 1,
    "ClusterDomain": "",
    "ClusterAlias": "my-cluster",
    "ClusterName": "data:3306"
  },
  {
    "HasAutomatedIntermediateMasterRecovery": true,
    "HasAutomatedMasterRecovery": true,
    "HeuristicLag": 0,
    "CountInstances": 3,
    "ClusterDomain": "",
    "ClusterAlias": "my-cluster",
    "ClusterName": "data2:3306"
  }
]
```

```
$ curl -n -s ${orchestrator_url}/api/masters | jq -r '.[].Key.Hostname'
data
data2
```

```
$ curl -n -s ${orchestrator_url}/api/master/my-cluster | jq -r '.Key.Hostname'
data
```

There are now two clusters in Orchestrator, and they share the same cluster alias. Why are there two clusters and not one (`data2:3306`) with a detached and offlined `data` instance?

Why does `data:3306` cluster still have the `my-cluster` alias? I assume this is due to the all-matching `ClusterNameToAlias`—if this is the case, should `data:3306` cluster not have an alias applied since its sole instance is dead?

Even though `data` is offline, it still shows up as the master of `my-cluster`. Why is that?

How can I determine the real master of `my-cluster` after recovery? `/api/masters` shows two masters, and `/master/my-cluster` shows the dead instance.

Thanks!

**Comment by shlomi-noach** [comment]
To begin with, can you please make sure to read https://github.com/github/orchestrator/blob/master/docs/configuration-discovery-classifying.md#cluster-alias? (also relevant is https://github.com/github/orchestrator/blob/master/docs/deployment.md#populating-meta-data)

I would suggest that you don't use `ClusterNameToAlias`. It is kept for backwards compatibility, and I strongly suggest you use `DetectClusterAliasQuery`.

But, since you've provided with excellent input (thank you!) here's the problem:
```
"ClusterNameToAlias": {
    ".*": "my-cluster"
  },
```

You have mapped all of the clusters into a single alias.

What you want to do is give a different logical name (the alias) to different clusters.

---

Now, the above does not fully answer your question. To complete the picture, and I'm happy if you can confirm: when `orchestrator` fails over a master, it marks the failed master as "downtimed", with a special `lost_in_recovery` message. Can you verify if that is the case?

You can:
```
orchestrator-client -c api -path cluster/my-cluster | jq -r '.[].IsDowntimed'
orchestrator-client -c api -path cluster/my-cluster | jq -r '.[].DowntimeReason'
```
or you can use the `curl`
```
curl -n -s ${orchestrator_url}/api/cluster/my-cluster | jq -r '.[].DowntimeReason'
```

If it's not downtimed, that's strange and let's investigate further. If it's downtimed, and says `lost_in_recovery`, then even if it says "hi, I'm my-cluster", the _other_ master, which is not downtimed with `lost_in_recovery`, will be recognized as the true master.

So, right now it seems like this isn't working for you -- but there's two problems together. Let's solve the first (alias) and proceed with the second, if you can be so kind as to reconfigure, retry, and re-supply the information.

**Comment by stankevich** [comment]
> To complete the picture, and I'm happy if you can confirm: when `orchestrator` fails over a master, it marks the failed master as "downtimed", with a special `lost_in_recovery` message. Can you verify if that is the case?

Verified on the above setup (with `ClusterNameToAlias` still in place). It doesn't look like the failed master is getting downtimed:

Before failover:

```
$ for instance in data data2 data3 data4; do curl -n -s ${orchestrator_url}/api/instance/${instance}/3306 | jq -r '{ IsDowntimed, DowntimeReason, IsLastCheckValid }'; done
{
  "IsLastCheckValid": true,
  "DowntimeReason": "",
  "IsDowntimed": false
}
{
  "IsLastCheckValid": true,
  "DowntimeReason": "",
  "IsDowntimed": false
}
{
  "IsLastCheckValid": true,
  "DowntimeReason": "",
  "IsDowntimed": false
}
{
  "IsLastCheckValid": true,
  "DowntimeReason": "",
  "IsDowntimed": false
}
```

After failover:

```
$ for i in data data2 data3 data4; do curl -n -s ${orchestrator_url}/api/instance/${instance}/3306 | jq -r '{ IsDowntimed, DowntimeReason, IsLastCheckValid }'; done
{
  "IsLastCheckValid": false,
  "DowntimeReason": "",
  "IsDowntimed": false
}
{
  "IsLastCheckValid": true,
  "DowntimeReason": "",
  "IsDowntimed": false
}
{
  "IsLastCheckValid": true,
  "DowntimeReason": "",
  "IsDowntimed": false
}
{
  "IsLastCheckValid": true,
  "DowntimeReason": "",
  "IsDowntimed": false
}
```

---

`ClusterNameToAlias` vs `DetectClusterAliasQuery` situation makes sense. If I understand it correctly, Orchestrator assigns a cluster alias only while MySQL is able to execute `DetectClusterAliasQuery` successfully, and falls back to `ClusterName` when `DetectClusterAliasQuery` fails (== MySQL is offline). Is this more or less correct?

I replaced `ClusterNameToAlias` with `DetectClusterAliasQuery: SELECT SUBSTRING_INDEX(SUBSTRING_INDEX(@@hostname, '.', 2), '.', -1)`. All instances are named as `data<n>.<cluster-name>.example.com`, so this query just extracts cluster alias from hostnames:

```
root@data.my-cluster.example.com [(none)]> SELECT SUBSTRING_INDEX(SUBSTRING_INDEX(@@hostname, '.', 2), '.', -1);
+---------------------------------------------------------------+
| SUBSTRING_INDEX(SUBSTRING_INDEX(@@hostname, '.', 2), '.', -1) |
+---------------------------------------------------------------+
| my-cluster                                                    |
+---------------------------------------------------------------+
1 row in set (0.00 sec)
```

Initial state:

```
$ curl -n -s ${orchestrator_url}/api/topology/my-cluster | jq -r '.Details'
data.my-cluster.example.com:3306    [0s,ok,5.7.20-19-log,rw,ROW,>>,GTID]
+ data2.my-cluster.example.com:3306 [0s,ok,5.7.20-19-log,ro,ROW,>>,GTID]
+ data3.my-cluster.example.com:3306 [0s,ok,5.7.20-19-log,ro,ROW,>>,GTID]
+ data4.my-cluster.example.com:3306 [0s,ok,5.7.20-19-log,ro,ROW,>>,GTID]
```

```
$ curl -n -s ${orchestrator_url}/api/clusters | jq -r .
[
  "data.my-cluster.example.com:3306"
]
```

```
$ curl -n -s ${orchestrator_url}/api/clusters-info | jq -r .
[
  {
    "HasAutomatedIntermediateMasterRecovery": true,
    "HasAutomatedMasterRecovery": true,
    "HeuristicLag": 0,
    "CountInstances": 4,
    "ClusterDomain": "",
    "ClusterAlias": "my-cluster",
    "ClusterName": "data.my-cluster.example.com:3306"
  }
]
```

```
$ curl -n -s ${orchestrator_url}/api/masters | jq -r '.[].Key.Hostname'
data.my-cluster.example.com
```

```
$ curl -n -s ${orchestrator_url}/api/master/my-cluster | jq -r '.Key.Hostname'
data.my-cluster.example.com
```

```
$ for instance in data data2 data3 data4; do curl -n -s ${orchestrator_url}/api/instance/${instance}.my-cluster.example.com/3306 | jq -r '{ IsDowntimed, DowntimeReason, IsLastCheckValid }'; done
{
  "IsLastCheckValid": true,
  "DowntimeReason": "",
  "IsDowntimed": false
}
{
  "IsLastCheckValid": true,
  "DowntimeReason": "",
  "IsDowntimed": false
}
{
  "IsLastCheckValid": true,
  "DowntimeReason": "",
  "IsDowntimed": false
}
{
  "IsLastCheckValid": true,
  "DowntimeReason": "",
  "IsDowntimed": false
}
```

After `data.my-cluster.example.com` fails and `data2.my-cluster.example.com` is promoted:

```
$ curl -n -s ${orchestrator_url}/api/topology/my-cluster | jq -r '.Details'
data2.my-cluster.example.com:3306   [0s,ok,5.7.20-19-log,rw,ROW,>>,GTID]
+ data3.my-cluster.example.com:3306 [0s,ok,5.7.20-19-log,ro,ROW,>>,GTID]
+ data4.my-cluster.example.com:3306 [0s,ok,5.7.20-19-log,ro,ROW,>>,GTID]
```

```
$ curl -n -s ${orchestrator_url}/api/clusters | jq -r .
[
  "data.my-cluster.example.com:3306",
  "data2.my-cluster.example.com:3306"
]
```

```
$ curl -n -s ${orchestrator_url}/api/clusters-info | jq -r .
[
  {
    "HasAutomatedIntermediateMasterRecovery": true,
    "HasAutomatedMasterRecovery": true,
    "HeuristicLag": 0,
    "CountInstances": 1,
    "ClusterDomain": "",
    "ClusterAlias": "data.my-cluster.example.com:3306",
    "ClusterName": "data.my-cluster.example.com:3306"
  },
  {
    "HasAutomatedIntermediateMasterRecovery": true,
    "HasAutomatedMasterRecovery": true,
    "HeuristicLag": 0,
    "CountInstances": 3,
    "ClusterDomain": "",
    "ClusterAlias": "my-cluster",
    "ClusterName": "data2.my-cluster.example.com:3306"
  }
]
```

```
$ curl -n -s ${orchestrator_url}/api/masters | jq -r '.[].Key.Hostname'
data.my-cluster.example.com
data2.my-cluster.example.com
```

```
$ curl -n -s ${orchestrator_url}/api/master/my-cluster | jq -r '.Key.Hostname'
data2.my-cluster.example.com
```

This is already looking better. `data.my-cluster.example.com` is no longer in `my-cluster`, and `/api/master/my-cluster` responds with the right master. However, failed master is still not downtimed for some reason:

```
$ for instance in data data2 data3 data4; do curl -n -s ${orchestrator_url}/api/instance/${instance}.my-cluster.example.com/3306 | jq -r '{ IsDowntimed, DowntimeReason, IsLastCheckValid }'; done
{
  "IsLastCheckValid": false,
  "DowntimeReason": "",
  "IsDowntimed": false
}
{
  "IsLastCheckValid": true,
  "DowntimeReason": "",
  "IsDowntimed": false
}
{
  "IsLastCheckValid": true,
  "DowntimeReason": "",
  "IsDowntimed": false
}
{
  "IsLastCheckValid": true,
  "DowntimeReason": "",
  "IsDowntimed": false
}
```

Another thing: after `data.my-cluster.example.com` fails, it's displayed as a separate cluster in web UI. However, clicking on the alias (`data.my-cluster.example.com:3306`) takes me to an error page (at https://orchestrator.example.com/web/cluster/alias/data.my-cluster.example.com%3A3306):

```json
{"Code":"ERROR","Message":"No cluster found for alias data.my-cluster.example.com:3306","Details":null}
```

Clicking on the instance name (below alias) works fine: https://orchestrator.example.com/web/cluster/data.my-cluster.example.com:3306

---

I guess I got confused by the fact that a cluster in Orchestrator is not a static, pre-defined number of DB instances but a representation of a healthy replication topology. Cluster is always named after current master; failed masters fall out of the cluster and form their own clusters. Am I missing something here?

Looks like my initial problem of determining the right master is solved by switching from the all-matching `ClusterNameToAlias` to `DetectClusterAliasQuery`.

I'd still like to figure out why failed master is not being downtimed by Orchestrator. Any ideas?

**Comment by shlomi-noach** [comment]
> If I understand it correctly, Orchestrator assigns a cluster alias only while MySQL is able to execute DetectClusterAliasQuery successfully, and falls back to ClusterName when DetectClusterAliasQuery fails (== MySQL is offline). Is this more or less correct?

Not exactly. Once it's set, and until told otherwise, that's the alias. MySQL can go down and still `orchestrator` will associate it with that alias.

Your `DetectClusterAliasQuery` looks excellent.

As for the rest-- I'm taking off for a long weekend in FOSDEM, so I can back to investigating your issue mid-next week.

**Comment by stankevich** [comment]
@shlomi-noach: Orchestrator is still not downtiming the failed master for me. Any suggestions?

**Comment by shlomi-noach** [comment]
Thanks for the reminder!

**Comment by shlomi-noach** [comment]
I'm still unsure why your failed master was not dontimed. It is expected to be so at https://github.com/github/orchestrator/blob/8a0db2b895fb527ced7c51a2cac001ff749c2d77/go/logic/topology_recovery.go#L537

I'm mereley verifying now that the `orchestrator` service is up and running throughout the operation, is not being killed; is that correct? (As opposed to, say, the `orchestrator` service not running at all, and all ops are done via the command-line client)

Immediately following the failover, can you query the backed:

```
select * from database_instance_downtime;
```

and possibly paste the `orchestrator` log starting at time of failover, ending after failover is complete?

**Comment by stankevich** [comment]
> I'm mereley verifying now that the orchestrator service is up and running throughout the operation, is not being killed; is that correct?

That is correct, Orchestrator continues to run just fine.

I queried the `database_instance_downtime` table on raft leader during failover: 

```
$ while true; do echo "$(date +'%T.%3N') $(sqlite3 -batch orchestrator.sqlite3 'select * from database_instance_downtime;')" | tee -a database_instance_downtime.log; sleep 0.3; done
[...]
12:21:09.350
12:21:09.658
12:21:09.965
12:21:10.273
12:21:10.580
12:21:10.915
12:21:11.223
12:21:11.530
12:21:11.837
12:21:12.145
12:21:12.453 data.my-cluster.example.com|3306|1|2018-02-14 12:21:12|2018-02-14 12:21:12|orchestrator.my-cluster.example.com|lost-in-recovery
12:21:12.771 data.my-cluster.example.com|3306|1|2018-02-14 12:21:12|2018-02-14 12:21:12|orchestrator.my-cluster.example.com|lost-in-recovery
12:21:13.078 data.my-cluster.example.com|3306|1|2018-02-14 12:21:12|2018-02-14 12:21:12|orchestrator.my-cluster.example.com|lost-in-recovery
12:21:13.385 data.my-cluster.example.com|3306|1|2018-02-14 12:21:12|2018-02-14 12:21:12|orchestrator.my-cluster.example.com|lost-in-recovery
12:21:13.693 data.my-cluster.example.com|3306|1|2018-02-14 12:21:12|2018-02-14 12:21:12|orchestrator.my-cluster.example.com|lost-in-recovery
12:21:14.002 data.my-cluster.example.com|3306|1|2018-02-14 12:21:12|2018-02-14 12:21:12|orchestrator.my-cluster.example.com|lost-in-recovery
12:21:14.321 data.my-cluster.example.com|3306|1|2018-02-14 12:21:12|2018-02-14 12:21:12|orchestrator.my-cluster.example.com|lost-in-recovery
12:21:14.629 data.my-cluster.example.com|3306|1|2018-02-14 12:21:12|2018-02-14 12:21:12|orchestrator.my-cluster.example.com|lost-in-recovery
12:21:14.936 data.my-cluster.example.com|3306|1|2018-02-14 12:21:12|2018-02-14 12:21:12|orchestrator.my-cluster.example.com|lost-in-recovery
12:21:15.244
12:21:15.551
12:21:15.859
12:21:16.166
12:21:16.473
12:21:16.781
12:21:17.102
[...]
```

Looks like Orchestrator _is_ downtiming the failed master, but only briefly.

Debug log from this event: https://gist.github.com/stankevich/35edb788c4d320547097c1ec7037ad12

**Comment by shlomi-noach** [comment]
Aha!! I do believe https://github.com/github/orchestrator/pull/411 will fix this issue.

**Comment by shlomi-noach** [comment]
Released in [3.0.7](https://github.com/github/orchestrator/releases/latest)
